{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "240f5970",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import io\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import boto3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6d4b771",
   "metadata": {},
   "outputs": [],
   "source": [
    "session = boto3.Session(profile_name='insight')\n",
    "athena_client = session.client('athena')\n",
    "s3_client = session.client('s3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2958f1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wait_for_query_completion(client, query_execution_id: str):\n",
    "    \"\"\"      \n",
    "    determines if the query has finished excuting. If it has not, the function waits for the query to finished.\n",
    "    Parameters:\n",
    "        query_execution_id: the query execution id (as a string)\n",
    "        client: an aws client or session.client\n",
    "        returns the location of query results (string)\n",
    "    \"\"\"\n",
    "    max_execution = 600\n",
    "    state = 'RUNNING'\n",
    "    while (max_execution > 0 and state in ['RUNNING', 'QUEUED']):\n",
    "        max_execution = max_execution - 1\n",
    "        response = client.get_query_execution(QueryExecutionId = query_execution_id)\n",
    "        if 'QueryExecution' in response and 'Status' in response['QueryExecution'] and 'State' in response['QueryExecution']['Status']:\n",
    "            state = response['QueryExecution']['Status']['State']\n",
    "            if state == 'FAILED':\n",
    "                return state\n",
    "            elif state == 'SUCCEEDED':\n",
    "                s3_path = response['QueryExecution']['ResultConfiguration']['OutputLocation']\n",
    "                return s3_path\n",
    "        time.sleep(1)\n",
    "    return 'FAILED'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec25f868",
   "metadata": {},
   "outputs": [],
   "source": [
    "context = {\n",
    "    'Database': 'genes_and_health',\n",
    "    'Catalog': 'AwsDataCatalog'\n",
    "}\n",
    "local_query_string = 'SELECT * FROM qt_gene_based LIMIT 50;'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7e80db8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 18.9 ms, sys: 2.79 ms, total: 21.7 ms\n",
      "Wall time: 3.23 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "response = athena_client.start_query_execution(QueryString=local_query_string, QueryExecutionContext=context)\n",
    "query_execution_id = response['QueryExecutionId']\n",
    "#print('Execution ID: ', query_execution_id)\n",
    "s3_path = wait_for_query_completion(athena_client, query_execution_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b748f66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s3://aws-athena-query-results-638341270768-us-east-1/ef58b7b7-2621-4a8e-9293-d74213e802ce.csv'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s3_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9a3a31f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file name:  ef58b7b7-2621-4a8e-9293-d74213e802ce.csv file location:  aws-athena-query-results-638341270768-us-east-1\n"
     ]
    }
   ],
   "source": [
    "file_name = re.findall('.*\\/(.*)', s3_path)[0]\n",
    "file_location = s3_path.replace('s3://','').split('/', 1)[0]\n",
    "print('file name: ', file_name, 'file location: ', file_location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5017bfda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>genpos</th>\n",
       "      <th>id</th>\n",
       "      <th>mask</th>\n",
       "      <th>a1freq</th>\n",
       "      <th>n</th>\n",
       "      <th>test</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>chisq</th>\n",
       "      <th>...</th>\n",
       "      <th>gene id</th>\n",
       "      <th>variant mask</th>\n",
       "      <th>freq</th>\n",
       "      <th>raw path</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>stat</th>\n",
       "      <th>gene-based</th>\n",
       "      <th>female-only</th>\n",
       "      <th>male-only</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.singleton</td>\n",
       "      <td>B.singleton</td>\n",
       "      <td>0.003036</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.363867</td>\n",
       "      <td>0.353264</td>\n",
       "      <td>1.060920</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.0.0001</td>\n",
       "      <td>B.0.0001</td>\n",
       "      <td>0.003036</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.363867</td>\n",
       "      <td>0.353264</td>\n",
       "      <td>1.060920</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.074253</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.074253</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>0.003903</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.085238</td>\n",
       "      <td>0.312806</td>\n",
       "      <td>0.074253</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_B.0.01</td>\n",
       "      <td>B.0.01</td>\n",
       "      <td>0.009974</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.055619</td>\n",
       "      <td>0.187779</td>\n",
       "      <td>0.087732</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>B</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.singleton</td>\n",
       "      <td>C.singleton</td>\n",
       "      <td>0.007372</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.328905</td>\n",
       "      <td>0.227895</td>\n",
       "      <td>2.082920</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.0.0001</td>\n",
       "      <td>C.0.0001</td>\n",
       "      <td>0.007372</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.328905</td>\n",
       "      <td>0.227895</td>\n",
       "      <td>2.082920</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.646439</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.646439</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>0.012576</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.141121</td>\n",
       "      <td>0.175520</td>\n",
       "      <td>0.646439</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_C.0.01</td>\n",
       "      <td>C.0.01</td>\n",
       "      <td>0.033391</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.020931</td>\n",
       "      <td>0.108133</td>\n",
       "      <td>0.037469</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>C</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.singleton</td>\n",
       "      <td>D.singleton</td>\n",
       "      <td>0.006071</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.087428</td>\n",
       "      <td>0.251327</td>\n",
       "      <td>0.121010</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.0.0001</td>\n",
       "      <td>D.0.0001</td>\n",
       "      <td>0.006071</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.087428</td>\n",
       "      <td>0.251327</td>\n",
       "      <td>0.121010</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.267154</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.267154</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>0.008673</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.108981</td>\n",
       "      <td>0.210848</td>\n",
       "      <td>0.267154</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>925926</td>\n",
       "      <td>SAMD11(ENSG00000187634).MASK_D.0.01</td>\n",
       "      <td>D.0.01</td>\n",
       "      <td>0.021249</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.018623</td>\n",
       "      <td>0.136802</td>\n",
       "      <td>0.018531</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>D</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.singleton</td>\n",
       "      <td>B.singleton</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.421493</td>\n",
       "      <td>0.417472</td>\n",
       "      <td>1.019360</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.0.0001</td>\n",
       "      <td>B.0.0001</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.421493</td>\n",
       "      <td>0.417472</td>\n",
       "      <td>1.019360</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.019360</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.019360</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.421493</td>\n",
       "      <td>0.417472</td>\n",
       "      <td>1.019360</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_B.0.01</td>\n",
       "      <td>B.0.01</td>\n",
       "      <td>0.003469</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.613385</td>\n",
       "      <td>0.329567</td>\n",
       "      <td>3.464010</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>B</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.singleton</td>\n",
       "      <td>C.singleton</td>\n",
       "      <td>0.009974</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.160581</td>\n",
       "      <td>0.196926</td>\n",
       "      <td>0.664935</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.0.0001</td>\n",
       "      <td>C.0.0001</td>\n",
       "      <td>0.009974</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.160581</td>\n",
       "      <td>0.196926</td>\n",
       "      <td>0.664935</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.347961</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.347961</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>0.012576</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.103575</td>\n",
       "      <td>0.175586</td>\n",
       "      <td>0.347961</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_C.0.01</td>\n",
       "      <td>C.0.01</td>\n",
       "      <td>0.027320</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.006830</td>\n",
       "      <td>0.119629</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>C</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.singleton</td>\n",
       "      <td>D.singleton</td>\n",
       "      <td>0.003469</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.221235</td>\n",
       "      <td>0.331840</td>\n",
       "      <td>0.444481</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.0.0001</td>\n",
       "      <td>D.0.0001</td>\n",
       "      <td>0.003469</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.221235</td>\n",
       "      <td>0.331840</td>\n",
       "      <td>0.444481</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.047461</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.047461</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>0.004337</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.064655</td>\n",
       "      <td>0.296781</td>\n",
       "      <td>0.047461</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>944699</td>\n",
       "      <td>NOC2L(ENSG00000188976).MASK_D.0.01</td>\n",
       "      <td>D.0.01</td>\n",
       "      <td>0.035559</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.041899</td>\n",
       "      <td>0.107995</td>\n",
       "      <td>0.150526</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>D</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.singleton</td>\n",
       "      <td>B.singleton</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.149986</td>\n",
       "      <td>0.419297</td>\n",
       "      <td>0.127954</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.0.0001</td>\n",
       "      <td>B.0.0001</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.149986</td>\n",
       "      <td>0.419297</td>\n",
       "      <td>0.127954</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.384130</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.384130</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>0.003903</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.194008</td>\n",
       "      <td>0.313026</td>\n",
       "      <td>0.384130</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_B.0.01</td>\n",
       "      <td>B.0.01</td>\n",
       "      <td>0.006938</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.035990</td>\n",
       "      <td>0.235605</td>\n",
       "      <td>0.023334</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>B</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.singleton</td>\n",
       "      <td>C.singleton</td>\n",
       "      <td>0.004770</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.156149</td>\n",
       "      <td>0.282833</td>\n",
       "      <td>0.304802</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.0.0001</td>\n",
       "      <td>C.0.0001</td>\n",
       "      <td>0.004770</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.156149</td>\n",
       "      <td>0.282833</td>\n",
       "      <td>0.304802</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011171</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011171</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.0.001</td>\n",
       "      <td>C.0.001</td>\n",
       "      <td>0.008239</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.216315</td>\n",
       "      <td>0.011171</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_C.0.01</td>\n",
       "      <td>C.0.01</td>\n",
       "      <td>0.015611</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.027623</td>\n",
       "      <td>0.154264</td>\n",
       "      <td>0.032063</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>C</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_D.singleton</td>\n",
       "      <td>D.singleton</td>\n",
       "      <td>0.005204</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.039680</td>\n",
       "      <td>0.270492</td>\n",
       "      <td>0.021520</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>D</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1</td>\n",
       "      <td>961299</td>\n",
       "      <td>KLHL17(ENSG00000187961).MASK_D.0.0001</td>\n",
       "      <td>D.0.0001</td>\n",
       "      <td>0.005204</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.039680</td>\n",
       "      <td>0.270492</td>\n",
       "      <td>0.021520</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom  genpos                                        id         mask  \\\n",
       "0       1  925926  SAMD11(ENSG00000187634).MASK_B.singleton  B.singleton   \n",
       "1       1  925926     SAMD11(ENSG00000187634).MASK_B.0.0001     B.0.0001   \n",
       "2       1  925926      SAMD11(ENSG00000187634).MASK_B.0.001      B.0.001   \n",
       "3       1  925926      SAMD11(ENSG00000187634).MASK_B.0.001      B.0.001   \n",
       "4       1  925926      SAMD11(ENSG00000187634).MASK_B.0.001      B.0.001   \n",
       "5       1  925926       SAMD11(ENSG00000187634).MASK_B.0.01       B.0.01   \n",
       "6       1  925926  SAMD11(ENSG00000187634).MASK_C.singleton  C.singleton   \n",
       "7       1  925926     SAMD11(ENSG00000187634).MASK_C.0.0001     C.0.0001   \n",
       "8       1  925926      SAMD11(ENSG00000187634).MASK_C.0.001      C.0.001   \n",
       "9       1  925926      SAMD11(ENSG00000187634).MASK_C.0.001      C.0.001   \n",
       "10      1  925926      SAMD11(ENSG00000187634).MASK_C.0.001      C.0.001   \n",
       "11      1  925926       SAMD11(ENSG00000187634).MASK_C.0.01       C.0.01   \n",
       "12      1  925926  SAMD11(ENSG00000187634).MASK_D.singleton  D.singleton   \n",
       "13      1  925926     SAMD11(ENSG00000187634).MASK_D.0.0001     D.0.0001   \n",
       "14      1  925926      SAMD11(ENSG00000187634).MASK_D.0.001      D.0.001   \n",
       "15      1  925926      SAMD11(ENSG00000187634).MASK_D.0.001      D.0.001   \n",
       "16      1  925926      SAMD11(ENSG00000187634).MASK_D.0.001      D.0.001   \n",
       "17      1  925926       SAMD11(ENSG00000187634).MASK_D.0.01       D.0.01   \n",
       "18      1  944699   NOC2L(ENSG00000188976).MASK_B.singleton  B.singleton   \n",
       "19      1  944699      NOC2L(ENSG00000188976).MASK_B.0.0001     B.0.0001   \n",
       "20      1  944699       NOC2L(ENSG00000188976).MASK_B.0.001      B.0.001   \n",
       "21      1  944699       NOC2L(ENSG00000188976).MASK_B.0.001      B.0.001   \n",
       "22      1  944699       NOC2L(ENSG00000188976).MASK_B.0.001      B.0.001   \n",
       "23      1  944699        NOC2L(ENSG00000188976).MASK_B.0.01       B.0.01   \n",
       "24      1  944699   NOC2L(ENSG00000188976).MASK_C.singleton  C.singleton   \n",
       "25      1  944699      NOC2L(ENSG00000188976).MASK_C.0.0001     C.0.0001   \n",
       "26      1  944699       NOC2L(ENSG00000188976).MASK_C.0.001      C.0.001   \n",
       "27      1  944699       NOC2L(ENSG00000188976).MASK_C.0.001      C.0.001   \n",
       "28      1  944699       NOC2L(ENSG00000188976).MASK_C.0.001      C.0.001   \n",
       "29      1  944699        NOC2L(ENSG00000188976).MASK_C.0.01       C.0.01   \n",
       "30      1  944699   NOC2L(ENSG00000188976).MASK_D.singleton  D.singleton   \n",
       "31      1  944699      NOC2L(ENSG00000188976).MASK_D.0.0001     D.0.0001   \n",
       "32      1  944699       NOC2L(ENSG00000188976).MASK_D.0.001      D.0.001   \n",
       "33      1  944699       NOC2L(ENSG00000188976).MASK_D.0.001      D.0.001   \n",
       "34      1  944699       NOC2L(ENSG00000188976).MASK_D.0.001      D.0.001   \n",
       "35      1  944699        NOC2L(ENSG00000188976).MASK_D.0.01       D.0.01   \n",
       "36      1  961299  KLHL17(ENSG00000187961).MASK_B.singleton  B.singleton   \n",
       "37      1  961299     KLHL17(ENSG00000187961).MASK_B.0.0001     B.0.0001   \n",
       "38      1  961299      KLHL17(ENSG00000187961).MASK_B.0.001      B.0.001   \n",
       "39      1  961299      KLHL17(ENSG00000187961).MASK_B.0.001      B.0.001   \n",
       "40      1  961299      KLHL17(ENSG00000187961).MASK_B.0.001      B.0.001   \n",
       "41      1  961299       KLHL17(ENSG00000187961).MASK_B.0.01       B.0.01   \n",
       "42      1  961299  KLHL17(ENSG00000187961).MASK_C.singleton  C.singleton   \n",
       "43      1  961299     KLHL17(ENSG00000187961).MASK_C.0.0001     C.0.0001   \n",
       "44      1  961299      KLHL17(ENSG00000187961).MASK_C.0.001      C.0.001   \n",
       "45      1  961299      KLHL17(ENSG00000187961).MASK_C.0.001      C.0.001   \n",
       "46      1  961299      KLHL17(ENSG00000187961).MASK_C.0.001      C.0.001   \n",
       "47      1  961299       KLHL17(ENSG00000187961).MASK_C.0.01       C.0.01   \n",
       "48      1  961299  KLHL17(ENSG00000187961).MASK_D.singleton  D.singleton   \n",
       "49      1  961299     KLHL17(ENSG00000187961).MASK_D.0.0001     D.0.0001   \n",
       "\n",
       "      a1freq     n       test      beta        se     chisq  ...  \\\n",
       "0   0.003036  1153        ADD  0.363867  0.353264  1.060920  ...   \n",
       "1   0.003036  1153        ADD  0.363867  0.353264  1.060920  ...   \n",
       "2        NaN  1153   ADD-SKAT       NaN       NaN  0.074253  ...   \n",
       "3        NaN  1153  ADD-SKATO       NaN       NaN  0.074253  ...   \n",
       "4   0.003903  1153        ADD  0.085238  0.312806  0.074253  ...   \n",
       "5   0.009974  1153        ADD -0.055619  0.187779  0.087732  ...   \n",
       "6   0.007372  1153        ADD  0.328905  0.227895  2.082920  ...   \n",
       "7   0.007372  1153        ADD  0.328905  0.227895  2.082920  ...   \n",
       "8        NaN  1153   ADD-SKAT       NaN       NaN  0.646439  ...   \n",
       "9        NaN  1153  ADD-SKATO       NaN       NaN  0.646439  ...   \n",
       "10  0.012576  1153        ADD  0.141121  0.175520  0.646439  ...   \n",
       "11  0.033391  1153        ADD  0.020931  0.108133  0.037469  ...   \n",
       "12  0.006071  1153        ADD  0.087428  0.251327  0.121010  ...   \n",
       "13  0.006071  1153        ADD  0.087428  0.251327  0.121010  ...   \n",
       "14       NaN  1153   ADD-SKAT       NaN       NaN  0.267154  ...   \n",
       "15       NaN  1153  ADD-SKATO       NaN       NaN  0.267154  ...   \n",
       "16  0.008673  1153        ADD -0.108981  0.210848  0.267154  ...   \n",
       "17  0.021249  1153        ADD  0.018623  0.136802  0.018531  ...   \n",
       "18  0.002168  1153        ADD -0.421493  0.417472  1.019360  ...   \n",
       "19  0.002168  1153        ADD -0.421493  0.417472  1.019360  ...   \n",
       "20       NaN  1153   ADD-SKAT       NaN       NaN  1.019360  ...   \n",
       "21       NaN  1153  ADD-SKATO       NaN       NaN  1.019360  ...   \n",
       "22  0.002168  1153        ADD -0.421493  0.417472  1.019360  ...   \n",
       "23  0.003469  1153        ADD -0.613385  0.329567  3.464010  ...   \n",
       "24  0.009974  1153        ADD -0.160581  0.196926  0.664935  ...   \n",
       "25  0.009974  1153        ADD -0.160581  0.196926  0.664935  ...   \n",
       "26       NaN  1153   ADD-SKAT       NaN       NaN  0.347961  ...   \n",
       "27       NaN  1153  ADD-SKATO       NaN       NaN  0.347961  ...   \n",
       "28  0.012576  1153        ADD -0.103575  0.175586  0.347961  ...   \n",
       "29  0.027320  1153        ADD -0.006830  0.119629  0.003259  ...   \n",
       "30  0.003469  1153        ADD -0.221235  0.331840  0.444481  ...   \n",
       "31  0.003469  1153        ADD -0.221235  0.331840  0.444481  ...   \n",
       "32       NaN  1153   ADD-SKAT       NaN       NaN  0.047461  ...   \n",
       "33       NaN  1153  ADD-SKATO       NaN       NaN  0.047461  ...   \n",
       "34  0.004337  1153        ADD -0.064655  0.296781  0.047461  ...   \n",
       "35  0.035559  1153        ADD  0.041899  0.107995  0.150526  ...   \n",
       "36  0.002168  1153        ADD  0.149986  0.419297  0.127954  ...   \n",
       "37  0.002168  1153        ADD  0.149986  0.419297  0.127954  ...   \n",
       "38       NaN  1153   ADD-SKAT       NaN       NaN  0.384130  ...   \n",
       "39       NaN  1153  ADD-SKATO       NaN       NaN  0.384130  ...   \n",
       "40  0.003903  1153        ADD  0.194008  0.313026  0.384130  ...   \n",
       "41  0.006938  1153        ADD  0.035990  0.235605  0.023334  ...   \n",
       "42  0.004770  1153        ADD  0.156149  0.282833  0.304802  ...   \n",
       "43  0.004770  1153        ADD  0.156149  0.282833  0.304802  ...   \n",
       "44       NaN  1153   ADD-SKAT       NaN       NaN  0.011171  ...   \n",
       "45       NaN  1153  ADD-SKATO       NaN       NaN  0.011171  ...   \n",
       "46  0.008239  1153        ADD  0.022863  0.216315  0.011171  ...   \n",
       "47  0.015611  1153        ADD -0.027623  0.154264  0.032063  ...   \n",
       "48  0.005204  1153        ADD -0.039680  0.270492  0.021520  ...   \n",
       "49  0.005204  1153        ADD -0.039680  0.270492  0.021520  ...   \n",
       "\n",
       "            gene id  variant mask       freq  \\\n",
       "0   ENSG00000187634             B  singleton   \n",
       "1   ENSG00000187634             B     0.0001   \n",
       "2   ENSG00000187634             B      0.001   \n",
       "3   ENSG00000187634             B      0.001   \n",
       "4   ENSG00000187634             B      0.001   \n",
       "5   ENSG00000187634             B       0.01   \n",
       "6   ENSG00000187634             C  singleton   \n",
       "7   ENSG00000187634             C     0.0001   \n",
       "8   ENSG00000187634             C      0.001   \n",
       "9   ENSG00000187634             C      0.001   \n",
       "10  ENSG00000187634             C      0.001   \n",
       "11  ENSG00000187634             C       0.01   \n",
       "12  ENSG00000187634             D  singleton   \n",
       "13  ENSG00000187634             D     0.0001   \n",
       "14  ENSG00000187634             D      0.001   \n",
       "15  ENSG00000187634             D      0.001   \n",
       "16  ENSG00000187634             D      0.001   \n",
       "17  ENSG00000187634             D       0.01   \n",
       "18  ENSG00000188976             B  singleton   \n",
       "19  ENSG00000188976             B     0.0001   \n",
       "20  ENSG00000188976             B      0.001   \n",
       "21  ENSG00000188976             B      0.001   \n",
       "22  ENSG00000188976             B      0.001   \n",
       "23  ENSG00000188976             B       0.01   \n",
       "24  ENSG00000188976             C  singleton   \n",
       "25  ENSG00000188976             C     0.0001   \n",
       "26  ENSG00000188976             C      0.001   \n",
       "27  ENSG00000188976             C      0.001   \n",
       "28  ENSG00000188976             C      0.001   \n",
       "29  ENSG00000188976             C       0.01   \n",
       "30  ENSG00000188976             D  singleton   \n",
       "31  ENSG00000188976             D     0.0001   \n",
       "32  ENSG00000188976             D      0.001   \n",
       "33  ENSG00000188976             D      0.001   \n",
       "34  ENSG00000188976             D      0.001   \n",
       "35  ENSG00000188976             D       0.01   \n",
       "36  ENSG00000187961             B  singleton   \n",
       "37  ENSG00000187961             B     0.0001   \n",
       "38  ENSG00000187961             B      0.001   \n",
       "39  ENSG00000187961             B      0.001   \n",
       "40  ENSG00000187961             B      0.001   \n",
       "41  ENSG00000187961             B       0.01   \n",
       "42  ENSG00000187961             C  singleton   \n",
       "43  ENSG00000187961             C     0.0001   \n",
       "44  ENSG00000187961             C      0.001   \n",
       "45  ENSG00000187961             C      0.001   \n",
       "46  ENSG00000187961             C      0.001   \n",
       "47  ENSG00000187961             C       0.01   \n",
       "48  ENSG00000187961             D  singleton   \n",
       "49  ENSG00000187961             D     0.0001   \n",
       "\n",
       "                                             raw path phenotype stat  \\\n",
       "0   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "1   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "2   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "3   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "4   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "5   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "6   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "7   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "8   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "9   tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "10  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "11  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "12  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "13  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "14  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "15  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "16  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "17  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "18  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "19  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "20  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "21  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "22  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "23  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "24  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "25  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "26  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "27  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "28  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "29  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "30  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "31  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "32  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "33  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "34  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "35  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "36  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "37  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "38  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "39  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "40  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "41  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "42  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "43  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "44  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "45  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "46  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "47  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "48  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "49  tak-insight-open-geneshealthconsort-silv/raw/2...       AFP  max   \n",
       "\n",
       "   gene-based female-only male-only         label  \n",
       "0        True       False     False  AFP (burden)  \n",
       "1        True       False     False  AFP (burden)  \n",
       "2        True       False     False  AFP (burden)  \n",
       "3        True       False     False  AFP (burden)  \n",
       "4        True       False     False  AFP (burden)  \n",
       "5        True       False     False  AFP (burden)  \n",
       "6        True       False     False  AFP (burden)  \n",
       "7        True       False     False  AFP (burden)  \n",
       "8        True       False     False  AFP (burden)  \n",
       "9        True       False     False  AFP (burden)  \n",
       "10       True       False     False  AFP (burden)  \n",
       "11       True       False     False  AFP (burden)  \n",
       "12       True       False     False  AFP (burden)  \n",
       "13       True       False     False  AFP (burden)  \n",
       "14       True       False     False  AFP (burden)  \n",
       "15       True       False     False  AFP (burden)  \n",
       "16       True       False     False  AFP (burden)  \n",
       "17       True       False     False  AFP (burden)  \n",
       "18       True       False     False  AFP (burden)  \n",
       "19       True       False     False  AFP (burden)  \n",
       "20       True       False     False  AFP (burden)  \n",
       "21       True       False     False  AFP (burden)  \n",
       "22       True       False     False  AFP (burden)  \n",
       "23       True       False     False  AFP (burden)  \n",
       "24       True       False     False  AFP (burden)  \n",
       "25       True       False     False  AFP (burden)  \n",
       "26       True       False     False  AFP (burden)  \n",
       "27       True       False     False  AFP (burden)  \n",
       "28       True       False     False  AFP (burden)  \n",
       "29       True       False     False  AFP (burden)  \n",
       "30       True       False     False  AFP (burden)  \n",
       "31       True       False     False  AFP (burden)  \n",
       "32       True       False     False  AFP (burden)  \n",
       "33       True       False     False  AFP (burden)  \n",
       "34       True       False     False  AFP (burden)  \n",
       "35       True       False     False  AFP (burden)  \n",
       "36       True       False     False  AFP (burden)  \n",
       "37       True       False     False  AFP (burden)  \n",
       "38       True       False     False  AFP (burden)  \n",
       "39       True       False     False  AFP (burden)  \n",
       "40       True       False     False  AFP (burden)  \n",
       "41       True       False     False  AFP (burden)  \n",
       "42       True       False     False  AFP (burden)  \n",
       "43       True       False     False  AFP (burden)  \n",
       "44       True       False     False  AFP (burden)  \n",
       "45       True       False     False  AFP (burden)  \n",
       "46       True       False     False  AFP (burden)  \n",
       "47       True       False     False  AFP (burden)  \n",
       "48       True       False     False  AFP (burden)  \n",
       "49       True       False     False  AFP (burden)  \n",
       "\n",
       "[50 rows x 23 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s3_client = session.client('s3')\n",
    "obj = s3_client.get_object(Bucket=file_location, Key=file_name)\n",
    "result_df = pd.read_csv(io.BytesIO(obj['Body'].read()), header=0)\n",
    "result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b191ae",
   "metadata": {},
   "source": [
    "## Query for all results for a given gene (TYK2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e7d02684",
   "metadata": {},
   "outputs": [],
   "source": [
    "queryStringGene = 'SELECT * FROM qt_gene_based WHERE qt_gene_based.\"gene id\"=\\'ENSG00000105397\\';'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ab18eb06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 616 ms, sys: 26.6 ms, total: 643 ms\n",
      "Wall time: 3min 20s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "response = athena_client.start_query_execution(QueryString=queryStringGene, QueryExecutionContext=context)\n",
    "query_execution_id = response['QueryExecutionId']\n",
    "s3_path = wait_for_query_completion(athena_client, query_execution_id)\n",
    "if not s3_path == 'FAILED':\n",
    "    file_name = re.findall('.*\\/(.*)', s3_path)[0]\n",
    "    file_location = s3_path.replace('s3://','').split('/', 1)[0]\n",
    "    obj = s3_client.get_object(Bucket=file_location, Key=file_name)\n",
    "    result_df = pd.read_csv(io.BytesIO(obj['Body'].read()), header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fe310cfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>genpos</th>\n",
       "      <th>id</th>\n",
       "      <th>mask</th>\n",
       "      <th>a1freq</th>\n",
       "      <th>n</th>\n",
       "      <th>test</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>chisq</th>\n",
       "      <th>...</th>\n",
       "      <th>gene id</th>\n",
       "      <th>variant mask</th>\n",
       "      <th>freq</th>\n",
       "      <th>raw path</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>stat</th>\n",
       "      <th>gene-based</th>\n",
       "      <th>female-only</th>\n",
       "      <th>male-only</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_B.singleton</td>\n",
       "      <td>B.singleton</td>\n",
       "      <td>0.006505</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.279340</td>\n",
       "      <td>0.251489</td>\n",
       "      <td>1.233760</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>B</td>\n",
       "      <td>singleton</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_B.0.0001</td>\n",
       "      <td>B.0.0001</td>\n",
       "      <td>0.006505</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.279340</td>\n",
       "      <td>0.251489</td>\n",
       "      <td>1.233760</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.245660</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.245660</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_B.0.001</td>\n",
       "      <td>B.0.001</td>\n",
       "      <td>0.008239</td>\n",
       "      <td>1153</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.110291</td>\n",
       "      <td>0.222521</td>\n",
       "      <td>0.245660</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>B</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AFP</td>\n",
       "      <td>max</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AFP (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7303</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_D.0.0001</td>\n",
       "      <td>D.0.0001</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.018873</td>\n",
       "      <td>0.120692</td>\n",
       "      <td>0.024452</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>D</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7304</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD-SKAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.409387</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7305</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD-SKATO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.227166</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7306</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_D.0.001</td>\n",
       "      <td>D.0.001</td>\n",
       "      <td>0.008647</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.041688</td>\n",
       "      <td>0.057690</td>\n",
       "      <td>0.522176</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>D</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference (burden)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7307</th>\n",
       "      <td>19</td>\n",
       "      <td>10350840</td>\n",
       "      <td>TYK2(ENSG00000105397).MASK_D.0.01</td>\n",
       "      <td>D.0.01</td>\n",
       "      <td>0.035158</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.009734</td>\n",
       "      <td>0.029621</td>\n",
       "      <td>0.107994</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>D</td>\n",
       "      <td>0.01</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference (burden)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7308 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    genpos                                      id         mask  \\\n",
       "0        19  10350840  TYK2(ENSG00000105397).MASK_B.singleton  B.singleton   \n",
       "1        19  10350840     TYK2(ENSG00000105397).MASK_B.0.0001     B.0.0001   \n",
       "2        19  10350840      TYK2(ENSG00000105397).MASK_B.0.001      B.0.001   \n",
       "3        19  10350840      TYK2(ENSG00000105397).MASK_B.0.001      B.0.001   \n",
       "4        19  10350840      TYK2(ENSG00000105397).MASK_B.0.001      B.0.001   \n",
       "...     ...       ...                                     ...          ...   \n",
       "7303     19  10350840     TYK2(ENSG00000105397).MASK_D.0.0001     D.0.0001   \n",
       "7304     19  10350840      TYK2(ENSG00000105397).MASK_D.0.001      D.0.001   \n",
       "7305     19  10350840      TYK2(ENSG00000105397).MASK_D.0.001      D.0.001   \n",
       "7306     19  10350840      TYK2(ENSG00000105397).MASK_D.0.001      D.0.001   \n",
       "7307     19  10350840       TYK2(ENSG00000105397).MASK_D.0.01       D.0.01   \n",
       "\n",
       "        a1freq      n       test      beta        se     chisq  ...  \\\n",
       "0     0.006505   1153        ADD -0.279340  0.251489  1.233760  ...   \n",
       "1     0.006505   1153        ADD -0.279340  0.251489  1.233760  ...   \n",
       "2          NaN   1153   ADD-SKAT       NaN       NaN  0.245660  ...   \n",
       "3          NaN   1153  ADD-SKATO       NaN       NaN  0.245660  ...   \n",
       "4     0.008239   1153        ADD -0.110291  0.222521  0.245660  ...   \n",
       "...        ...    ...        ...       ...       ...       ...  ...   \n",
       "7303  0.002020  15843        ADD  0.018873  0.120692  0.024452  ...   \n",
       "7304       NaN  15843   ADD-SKAT       NaN       NaN  0.409387  ...   \n",
       "7305       NaN  15843  ADD-SKATO       NaN       NaN  0.227166  ...   \n",
       "7306  0.008647  15843        ADD -0.041688  0.057690  0.522176  ...   \n",
       "7307  0.035158  15843        ADD -0.009734  0.029621  0.107994  ...   \n",
       "\n",
       "              gene id  variant mask       freq  \\\n",
       "0     ENSG00000105397             B  singleton   \n",
       "1     ENSG00000105397             B     0.0001   \n",
       "2     ENSG00000105397             B      0.001   \n",
       "3     ENSG00000105397             B      0.001   \n",
       "4     ENSG00000105397             B      0.001   \n",
       "...               ...           ...        ...   \n",
       "7303  ENSG00000105397             D     0.0001   \n",
       "7304  ENSG00000105397             D      0.001   \n",
       "7305  ENSG00000105397             D      0.001   \n",
       "7306  ENSG00000105397             D      0.001   \n",
       "7307  ENSG00000105397             D       0.01   \n",
       "\n",
       "                                               raw path            phenotype  \\\n",
       "0     tak-insight-open-geneshealthconsort-silv/raw/2...                  AFP   \n",
       "1     tak-insight-open-geneshealthconsort-silv/raw/2...                  AFP   \n",
       "2     tak-insight-open-geneshealthconsort-silv/raw/2...                  AFP   \n",
       "3     tak-insight-open-geneshealthconsort-silv/raw/2...                  AFP   \n",
       "4     tak-insight-open-geneshealthconsort-silv/raw/2...                  AFP   \n",
       "...                                                 ...                  ...   \n",
       "7303  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "7304  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "7305  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "7306  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "7307  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "\n",
       "     stat gene-based female-only male-only                         label  \n",
       "0     max       True       False     False                  AFP (burden)  \n",
       "1     max       True       False     False                  AFP (burden)  \n",
       "2     max       True       False     False                  AFP (burden)  \n",
       "3     max       True       False     False                  AFP (burden)  \n",
       "4     max       True       False     False                  AFP (burden)  \n",
       "...   ...        ...         ...       ...                           ...  \n",
       "7303  min       True       False     False  Waist circumference (burden)  \n",
       "7304  min       True       False     False  Waist circumference (burden)  \n",
       "7305  min       True       False     False  Waist circumference (burden)  \n",
       "7306  min       True       False     False  Waist circumference (burden)  \n",
       "7307  min       True       False     False  Waist circumference (burden)  \n",
       "\n",
       "[7308 rows x 23 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c981d5d",
   "metadata": {},
   "source": [
    "## Query for all signigicant burden associations for a given gene (SERPINA1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a014131",
   "metadata": {},
   "outputs": [],
   "source": [
    "queryStringGene = 'SELECT * FROM qt_gene_based WHERE qt_gene_based.\"gene id\"=\\'ENSG00000197249\\' AND qt_gene_based.significant = true;'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0ec61ce3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 390 ms, sys: 15.4 ms, total: 405 ms\n",
      "Wall time: 3min 30s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "response = athena_client.start_query_execution(QueryString=queryStringGene, QueryExecutionContext=context)\n",
    "query_execution_id = response['QueryExecutionId']\n",
    "s3_path = wait_for_query_completion(athena_client, query_execution_id)\n",
    "if not s3_path == 'FAILED':\n",
    "    file_name = re.findall('.*\\/(.*)', s3_path)[0]\n",
    "    file_location = s3_path.replace('s3://','').split('/', 1)[0]\n",
    "    obj = s3_client.get_object(Bucket=file_location, Key=file_name)\n",
    "    result_df = pd.read_csv(io.BytesIO(obj['Body'].read()), header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7e6e06c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>genpos</th>\n",
       "      <th>id</th>\n",
       "      <th>mask</th>\n",
       "      <th>a1freq</th>\n",
       "      <th>n</th>\n",
       "      <th>test</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>chisq</th>\n",
       "      <th>...</th>\n",
       "      <th>gene id</th>\n",
       "      <th>variant mask</th>\n",
       "      <th>freq</th>\n",
       "      <th>raw path</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>stat</th>\n",
       "      <th>gene-based</th>\n",
       "      <th>female-only</th>\n",
       "      <th>male-only</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, genpos, id, mask, a1freq, n, test, beta, se, chisq, log10p, significant, gene symbol, gene id, variant mask, freq, raw path, phenotype, stat, gene-based, female-only, male-only, label]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 23 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f075929",
   "metadata": {},
   "source": [
    "## Query for all single variant associations within the genomic region of a given gene (TYK2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2106dca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "queryStringGene = 'SELECT * FROM qt_single_variant WHERE qt_single_variant.chrom = 19 AND qt_single_variant.genpos < 10380608 AND qt_single_variant.genpos > 10350533;'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0a7e03a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.11 s, sys: 62.9 ms, total: 1.17 s\n",
      "Wall time: 8min 40s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "response = athena_client.start_query_execution(QueryString=queryStringGene, QueryExecutionContext=context)\n",
    "query_execution_id = response['QueryExecutionId']\n",
    "s3_path = wait_for_query_completion(athena_client, query_execution_id)\n",
    "if not s3_path == 'FAILED':\n",
    "    file_name = re.findall('.*\\/(.*)', s3_path)[0]\n",
    "    file_location = s3_path.replace('s3://','').split('/', 1)[0]\n",
    "    obj = s3_client.get_object(Bucket=file_location, Key=file_name)\n",
    "    result_df = pd.read_csv(io.BytesIO(obj['Body'].read()), header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d63c8288",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>genpos</th>\n",
       "      <th>id</th>\n",
       "      <th>allele0</th>\n",
       "      <th>allele1</th>\n",
       "      <th>a1freq</th>\n",
       "      <th>n</th>\n",
       "      <th>test</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>chisq</th>\n",
       "      <th>log10p</th>\n",
       "      <th>significant</th>\n",
       "      <th>raw path</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>stat</th>\n",
       "      <th>gene-based</th>\n",
       "      <th>female-only</th>\n",
       "      <th>male-only</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>10350843</td>\n",
       "      <td>chr19_10350843_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>13584</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.008030</td>\n",
       "      <td>0.125962</td>\n",
       "      <td>0.004064</td>\n",
       "      <td>0.022655</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AST</td>\n",
       "      <td>max</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>10350910</td>\n",
       "      <td>chr19_10350910_T_C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.017005</td>\n",
       "      <td>13584</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.037337</td>\n",
       "      <td>0.044208</td>\n",
       "      <td>0.713312</td>\n",
       "      <td>0.399739</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AST</td>\n",
       "      <td>max</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>10351069</td>\n",
       "      <td>chr19_10351069_C_T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>13584</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.029673</td>\n",
       "      <td>0.387217</td>\n",
       "      <td>0.005872</td>\n",
       "      <td>0.027373</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AST</td>\n",
       "      <td>max</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>10351070</td>\n",
       "      <td>chr19_10351070_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001031</td>\n",
       "      <td>13584</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.062226</td>\n",
       "      <td>0.179512</td>\n",
       "      <td>0.120158</td>\n",
       "      <td>0.137354</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AST</td>\n",
       "      <td>max</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>10351109</td>\n",
       "      <td>chr19_10351109_G_C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000994</td>\n",
       "      <td>13584</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.206436</td>\n",
       "      <td>0.176301</td>\n",
       "      <td>1.371080</td>\n",
       "      <td>0.616854</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>AST</td>\n",
       "      <td>max</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40063</th>\n",
       "      <td>19</td>\n",
       "      <td>10378350</td>\n",
       "      <td>chr19_10378350_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.437253</td>\n",
       "      <td>0.192884</td>\n",
       "      <td>5.138910</td>\n",
       "      <td>1.630870</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40064</th>\n",
       "      <td>19</td>\n",
       "      <td>10378363</td>\n",
       "      <td>chr19_10378363_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.005144</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.054369</td>\n",
       "      <td>0.075027</td>\n",
       "      <td>0.525128</td>\n",
       "      <td>0.329140</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40065</th>\n",
       "      <td>19</td>\n",
       "      <td>10378365</td>\n",
       "      <td>chr19_10378365_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001168</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.098673</td>\n",
       "      <td>0.158914</td>\n",
       "      <td>0.385544</td>\n",
       "      <td>0.271930</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40066</th>\n",
       "      <td>19</td>\n",
       "      <td>10378371</td>\n",
       "      <td>chr19_10378371_A_C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>-0.205617</td>\n",
       "      <td>0.257894</td>\n",
       "      <td>0.635678</td>\n",
       "      <td>0.371325</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40067</th>\n",
       "      <td>19</td>\n",
       "      <td>10378397</td>\n",
       "      <td>chr19_10378397_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>15843</td>\n",
       "      <td>ADD</td>\n",
       "      <td>0.138431</td>\n",
       "      <td>0.139336</td>\n",
       "      <td>0.987052</td>\n",
       "      <td>0.494221</td>\n",
       "      <td>False</td>\n",
       "      <td>tak-insight-open-geneshealthconsort-silv/raw/2...</td>\n",
       "      <td>Waist_circumference</td>\n",
       "      <td>min</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Waist circumference</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40068 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom    genpos                  id allele0 allele1    a1freq      n  \\\n",
       "0         19  10350843  chr19_10350843_G_A       G       A  0.002098  13584   \n",
       "1         19  10350910  chr19_10350910_T_C       T       C  0.017005  13584   \n",
       "2         19  10351069  chr19_10351069_C_T       C       T  0.000221  13584   \n",
       "3         19  10351070  chr19_10351070_G_A       G       A  0.001031  13584   \n",
       "4         19  10351109  chr19_10351109_G_C       G       C  0.000994  13584   \n",
       "...      ...       ...                 ...     ...     ...       ...    ...   \n",
       "40063     19  10378350  chr19_10378350_G_A       G       A  0.000726  15843   \n",
       "40064     19  10378363  chr19_10378363_A_G       A       G  0.005144  15843   \n",
       "40065     19  10378365  chr19_10378365_G_A       G       A  0.001168  15843   \n",
       "40066     19  10378371  chr19_10378371_A_C       A       C  0.000379  15843   \n",
       "40067     19  10378397  chr19_10378397_G_A       G       A  0.001515  15843   \n",
       "\n",
       "      test      beta        se     chisq    log10p  significant  \\\n",
       "0      ADD  0.008030  0.125962  0.004064  0.022655        False   \n",
       "1      ADD -0.037337  0.044208  0.713312  0.399739        False   \n",
       "2      ADD -0.029673  0.387217  0.005872  0.027373        False   \n",
       "3      ADD  0.062226  0.179512  0.120158  0.137354        False   \n",
       "4      ADD -0.206436  0.176301  1.371080  0.616854        False   \n",
       "...    ...       ...       ...       ...       ...          ...   \n",
       "40063  ADD -0.437253  0.192884  5.138910  1.630870        False   \n",
       "40064  ADD  0.054369  0.075027  0.525128  0.329140        False   \n",
       "40065  ADD  0.098673  0.158914  0.385544  0.271930        False   \n",
       "40066  ADD -0.205617  0.257894  0.635678  0.371325        False   \n",
       "40067  ADD  0.138431  0.139336  0.987052  0.494221        False   \n",
       "\n",
       "                                                raw path            phenotype  \\\n",
       "0      tak-insight-open-geneshealthconsort-silv/raw/2...                  AST   \n",
       "1      tak-insight-open-geneshealthconsort-silv/raw/2...                  AST   \n",
       "2      tak-insight-open-geneshealthconsort-silv/raw/2...                  AST   \n",
       "3      tak-insight-open-geneshealthconsort-silv/raw/2...                  AST   \n",
       "4      tak-insight-open-geneshealthconsort-silv/raw/2...                  AST   \n",
       "...                                                  ...                  ...   \n",
       "40063  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "40064  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "40065  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "40066  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "40067  tak-insight-open-geneshealthconsort-silv/raw/2...  Waist_circumference   \n",
       "\n",
       "      stat  gene-based  female-only  male-only                label  \n",
       "0      max       False        False      False                  AST  \n",
       "1      max       False        False      False                  AST  \n",
       "2      max       False        False      False                  AST  \n",
       "3      max       False        False      False                  AST  \n",
       "4      max       False        False      False                  AST  \n",
       "...    ...         ...          ...        ...                  ...  \n",
       "40063  min       False        False      False  Waist circumference  \n",
       "40064  min       False        False      False  Waist circumference  \n",
       "40065  min       False        False      False  Waist circumference  \n",
       "40066  min       False        False      False  Waist circumference  \n",
       "40067  min       False        False      False  Waist circumference  \n",
       "\n",
       "[40068 rows x 20 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e5ee682",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1001e317",
   "metadata": {},
   "outputs": [],
   "source": [
    "athena_client.close()\n",
    "s3_client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549dab1b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GSP",
   "language": "python",
   "name": "gsp_venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
