{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23e47f33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tiledb v0.23.2\n",
      "tiledb-vcf v0.26.2\n",
      "tiledb-cloud v0.11.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import tiledb\n",
    "import tiledb.cloud\n",
    "import tiledb.cloud.utilities\n",
    "import tiledbvcf\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "print(\n",
    "    f\"tiledb v{tiledb.version.version}\\n\"\n",
    "    f\"tiledb-vcf v{tiledbvcf.version}\\n\"\n",
    "    f\"tiledb-cloud v{tiledb.cloud.version.version}\\n\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e346eb9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = tiledb.Config()\n",
    "config[\"vfs.s3.aws_access_key_id\"] = \"AKIAZJIA2ZTYOUU2AHZB\"\n",
    "config[\"vfs.s3.aws_secret_access_key\"] = \"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\"\n",
    "config[\"vfs.s3.region\"] = \"us-east-1\"\n",
    "vfs = tiledb.VFS(config=config)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d09e8704",
   "metadata": {},
   "source": [
    "## Ingest Genes and Health qt_gene_based table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3ee6ecae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 38s, sys: 56.4 s, total: 4min 34s\n",
      "Wall time: 8min 13s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gene_table = \"s3://tak-insight-open-geneshealthconsort-silv/processed/20240924/tiledb/qt_gene_based.tsv.gz\"\n",
    "csv_file = vfs.open(gene_table)\n",
    "tiledb.from_csv(\n",
    "    \"GNH_2023_02_44kCallset_qt_gene_based\",\n",
    "    csv_file,\n",
    "    mode='ingest',\n",
    "    sparse = True,\n",
    "    compression = 'gzip',\n",
    "    sep = '\\t',\n",
    "    low_memory = False,\n",
    "    chunksize = 5000000,\n",
    "    tile = 100000,\n",
    "    index_dims = ['Gene ID','Phenotype'],\n",
    "    converters = {'A1FREQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'BETA' : lambda x: -1 if x == np.nan else x,\n",
    "                  'SE' : lambda x: -1 if x == np.nan else x,\n",
    "                  'CHISQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'LOG10P' : lambda x: -1 if x == np.nan else x,\n",
    "                  'Mask': lambda x: '' if x == np.nan else x,\n",
    "                  'Variant Mask': lambda x: '' if x == np.nan else x,\n",
    "                  'Freq': lambda x: '' if x == np.nan else x,\n",
    "                 },\n",
    "    #ctx=tiledb.Ctx(config)\n",
    ")\n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e67147f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 1s, sys: 47.8 s, total: 2min 49s\n",
      "Wall time: 2min 4s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GNH_2023_02_44kCallset_qt_gene_based'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.consolidate(\"GNH_2023_02_44kCallset_qt_gene_based\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "679a6987",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.5 ms, sys: 191 ms, total: 216 ms\n",
      "Wall time: 1.56 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.vacuum(\"GNH_2023_02_44kCallset_qt_gene_based\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80695a1b",
   "metadata": {},
   "source": [
    "## Ingest Genes and Health qt_single_variant table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af3a478f",
   "metadata": {},
   "outputs": [
    {
     "ename": "TileDBError",
     "evalue": "Array URI 'GNH_2023_02_44kCallset_qt_single_variant' already exists!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTileDBError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m<timed exec>:3\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/tiledb/lib/python3.10/site-packages/tiledb/dataframe_.py:869\u001b[0m, in \u001b[0;36mfrom_csv\u001b[0;34m(uri, csv_file, **kwargs)\u001b[0m\n\u001b[1;32m    866\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m tiledb\u001b[38;5;241m.\u001b[39mTileDBError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid mode specified (\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(mode))\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mappend\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m tiledb\u001b[38;5;241m.\u001b[39marray_exists(uri):\n\u001b[0;32m--> 869\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m tiledb\u001b[38;5;241m.\u001b[39mTileDBError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mArray URI \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m already exists!\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(uri))\n\u001b[1;32m    871\u001b[0m \u001b[38;5;66;03m# this is a pandas pass-through argument, do not pop!\u001b[39;00m\n\u001b[1;32m    872\u001b[0m chunksize \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchunksize\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "\u001b[0;31mTileDBError\u001b[0m: Array URI 'GNH_2023_02_44kCallset_qt_single_variant' already exists!"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "variant_table = \"s3://tak-insight-open-geneshealthconsort-silv/processed/20240924/tiledb/qt_single_variant.tsv.gz\"\n",
    "csv_file = vfs.open(variant_table)\n",
    "tiledb.from_csv(\n",
    "    \"GNH_2023_02_44kCallset_qt_single_variant\",\n",
    "    csv_file,\n",
    "    mode='ingest',\n",
    "    sparse = True,\n",
    "    compression = 'gzip',\n",
    "    sep = '\\t',\n",
    "    low_memory = False,\n",
    "    chunksize = 5000000,\n",
    "    tile = 100000,\n",
    "    index_dims = ['CHROM','GENPOS','Phenotype'],\n",
    "    converters = {'A1FREQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'BETA' : lambda x: -1 if x == np.nan else x,\n",
    "                  'SE' : lambda x: -1 if x == np.nan else x,\n",
    "                  'CHISQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'LOG10P' : lambda x: -1 if x == np.nan else x,\n",
    "                 },\n",
    "    #ctx=tiledb.Ctx(config)\n",
    ")\n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7095d823",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5min 12s, sys: 1min 47s, total: 7min\n",
      "Wall time: 4min 55s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GNH_2023_02_44kCallset_qt_single_variant'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.consolidate(\"GNH_2023_02_44kCallset_qt_single_variant\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "72ba58f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 29.7 ms, sys: 337 ms, total: 367 ms\n",
      "Wall time: 1.59 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.vacuum(\"GNH_2023_02_44kCallset_qt_single_variant\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a3c6363",
   "metadata": {},
   "source": [
    "## Ingest Genes and Health bin_gene_based table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c39dcf14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 43min 4s, sys: 13min 53s, total: 56min 57s\n",
      "Wall time: 1h 26min 8s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gene_table = \"s3://tak-insight-open-geneshealthconsort-silv/processed/20240924/tiledb/bin_gene_based.tsv.gz\"\n",
    "csv_file = vfs.open(gene_table)\n",
    "tiledb.from_csv(\n",
    "    \"GNH_2023_02_44kCallset_bin_gene_based\",\n",
    "    csv_file,\n",
    "    mode='ingest',\n",
    "    sparse = True,\n",
    "    compression = 'gzip',\n",
    "    sep = '\\t',\n",
    "    low_memory = False,\n",
    "    chunksize = 5000000,\n",
    "    tile = 100000,\n",
    "    index_dims = ['Gene ID','Phenotype'],\n",
    "    converters = {'A1FREQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'BETA' : lambda x: -1 if x == np.nan else x,\n",
    "                  'SE' : lambda x: -1 if x == np.nan else x,\n",
    "                  'CHISQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'LOG10P' : lambda x: -1 if x == np.nan else x,\n",
    "                  'Mask': lambda x: '' if x == np.nan else x,\n",
    "                  'Variant Mask': lambda x: '' if x == np.nan else x,\n",
    "                  'Freq': lambda x: '' if x == np.nan else x,\n",
    "                 },\n",
    "    #ctx=tiledb.Ctx(config)\n",
    ")\n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "82714321",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23min 18s, sys: 5min 32s, total: 28min 50s\n",
      "Wall time: 18min 32s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GNH_2023_02_44kCallset_bin_gene_based'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.consolidate(\"GNH_2023_02_44kCallset_bin_gene_based\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6aa130b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 558 ms, sys: 2.14 s, total: 2.7 s\n",
      "Wall time: 5.01 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.vacuum(\"GNH_2023_02_44kCallset_bin_gene_based\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a35d94",
   "metadata": {},
   "source": [
    "## Ingest Genes and Health bin_single_variant table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "43e5bae3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2h 32min 14s, sys: 52min 7s, total: 3h 24min 22s\n",
      "Wall time: 6h 10min 38s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "variant_table = \"s3://tak-insight-open-geneshealthconsort-silv/processed/20240924/tiledb/bin_single_variant.tsv.gz\"\n",
    "csv_file = vfs.open(variant_table)\n",
    "tiledb.from_csv(\n",
    "    \"GNH_2023_02_44kCallset_bin_single_variant\",\n",
    "    csv_file,\n",
    "    mode='ingest',\n",
    "    sparse = True,\n",
    "    compression = 'gzip',\n",
    "    sep = '\\t',\n",
    "    low_memory = False,\n",
    "    chunksize = 5000000,\n",
    "    tile = 100000,\n",
    "    index_dims = ['CHROM','GENPOS','Phenotype'],\n",
    "    converters = {'A1FREQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'BETA' : lambda x: -1 if x == np.nan else x,\n",
    "                  'SE' : lambda x: -1 if x == np.nan else x,\n",
    "                  'CHISQ' : lambda x: -1 if x == np.nan else x,\n",
    "                  'LOG10P' : lambda x: -1 if x == np.nan else x,\n",
    "                 },\n",
    "    #ctx=tiledb.Ctx(config)\n",
    ")\n",
    "csv_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "77165954",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1h 50min 37s, sys: 32min 17s, total: 2h 22min 54s\n",
      "Wall time: 1h 27min 52s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GNH_2023_02_44kCallset_bin_single_variant'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.consolidate(\"GNH_2023_02_44kCallset_bin_single_variant\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40e766d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 786 ms, sys: 5.19 s, total: 5.98 s\n",
      "Wall time: 11.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tiledb.vacuum(\"GNH_2023_02_44kCallset_bin_single_variant\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244eceed",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TileDB",
   "language": "python",
   "name": "tiledb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
